### PR TITLE
Add cid to the preprepare message

### PR DESCRIPTION
**Pull request by Liorgal28**
We made this change to include the CID field in the preprepare message as part of our work on block accumulation.
The preprepare CID is the CID of the batch that will ultimately be written to the DB.
Though every request inside the preprepare has its own CID, they all will be part of the same batch CID.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -3609,7 +3609,7 @@ void ReplicaImp::executeRequestsInPrePrepareMsg(concordUtils::SpanWrapper &paren
     } else {
       LOG_DEBUG(CNSUS, "Consensus reached");
     }
-
+    SCOPED_MDC("pp_msg_cid", ppMsg->getCid());
```
Comment (teoparvanov) [inline bftengine/src/bftengine/ReplicaImp.cpp:3612]: this should also be logged when creating the batch

File: bftengine/src/bftengine/messages/PrePrepareMsg.cpp [hunk 2]
```diff
@@ -86,7 +86,41 @@ PrePrepareMsg::PrePrepareMsg(ReplicaId sender,
...
+                             CommitPath firstPath,
+                             const concordUtils::SpanContext& spanContext,
+                             size_t size,
+                             const std::string& correlationID)
+    : MessageBase(sender,
+                  MsgCode::PrePrepare,
+                  spanContext.data().size(),
+                  (((size + sizeof(Header) + correlationID.size()) < maxMessageSize<PrePrepareMsg>())
+                       ? (size + sizeof(Header) + correlationID.size())
+                       : maxMessageSize<PrePrepareMsg>() - spanContext.data().size()))
```
Comment (teoparvanov) [inline bftengine/src/bftengine/messages/PrePrepareMsg.cpp:114]: same here - talk to Dima

File: bftengine/src/bftengine/messages/PrePrepareMsg.cpp [hunk 3]
```diff
@@ -75,8 +75,8 @@ PrePrepareMsg::PrePrepareMsg(ReplicaId sender,
     : MessageBase(sender,
                   MsgCode::PrePrepare,
                   spanContext.data().size(),
-                  (((size + sizeof(Header)) < maxMessageSize<PrePrepareMsg>())
-                       ? (size + sizeof(Header))
+                  (((size + sizeof(Header) + std::to_string(s).size()) < maxMessageSize<PrePrepareMsg>())
+                       ? (size + sizeof(Header) + std::to_string(s).size())
                        : maxMessageSize<PrePrepareMsg>() - spanContext.data().size()))
```
Comment (teoparvanov) [inline bftengine/src/bftengine/messages/PrePrepareMsg.cpp:80]: @Liorgal28 can you please consult with @f-squirrel on the correctness of this logic here?

File: bftengine/src/bftengine/messages/PrePrepareMsg.hpp [hunk 4]
```diff
@@ -68,6 +69,14 @@ class PrePrepareMsg : public MessageBase {
                 const concordUtils::SpanContext& spanContext,
                 size_t size);
 
+  PrePrepareMsg(ReplicaId sender,
+                ViewNum v,
+                SeqNum s,
+                CommitPath firstPath,
+                const concordUtils::SpanContext& spanContext,
+                size_t size,
+                const std::string& correlationID);
```
Comment (teoparvanov) [inline bftengine/src/bftengine/messages/PrePrepareMsg.hpp:78]: let's call this `batchCid` or `batch_cid` everywhere in this PR

**Review by teoparvanov — COMMENTED** [review]
LGTM at first glance!

File: bftengine/src/bftengine/messages/PrePrepareMsg.hpp [hunk 5]
```diff
@@ -34,6 +34,7 @@ class PrePrepareMsg : public MessageBase {
     ViewNum viewNum;
     SeqNum seqNum;
     uint16_t flags;
+    uint64_t batch_cid_length;
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/PrePrepareMsg.hpp:37]: Please follow the current naming policy: batch_cid_length -> batchCidLength

File: bftengine/src/bftengine/messages/PrePrepareMsg.hpp [hunk 6]
```diff
@@ -68,6 +69,14 @@ class PrePrepareMsg : public MessageBase {
                 const concordUtils::SpanContext& spanContext,
                 size_t size);
 
+  PrePrepareMsg(ReplicaId sender,
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/PrePrepareMsg.hpp:72]: Please don't define additional constructors unless it is absolutely unavoidable. They are hard to maintain. You can give a default value to cid and reuse the existing one.

File: bftengine/src/bftengine/messages/PrePrepareMsg.cpp (hunk 2 above)
Comment (f-squirrel) [reply bftengine/src/bftengine/messages/PrePrepareMsg.cpp:114]: @teoparvanov , I am fine with the change:)

File: bftengine/src/bftengine/messages/PrePrepareMsg.cpp [hunk 7]
```diff
@@ -75,8 +75,8 @@ PrePrepareMsg::PrePrepareMsg(ReplicaId sender,
     : MessageBase(sender,
                   MsgCode::PrePrepare,
                   spanContext.data().size(),
-                  (((size + sizeof(Header)) < maxMessageSize<PrePrepareMsg>())
-                       ? (size + sizeof(Header))
+                  (((size + sizeof(Header) + std::to_string(s).size()) < maxMessageSize<PrePrepareMsg>())
```
Comment (f-squirrel) [inline bftengine/src/bftengine/messages/PrePrepareMsg.cpp:78]: Call the new `PrePrepareMsg` ctor.

**Review by f-squirrel — COMMENTED** [review]
LGTM, only small comments

File: bftengine/src/bftengine/messages/PrePrepareMsg.cpp (hunk 7 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/messages/PrePrepareMsg.cpp]: Done.

File: bftengine/src/bftengine/messages/PrePrepareMsg.hpp (hunk 6 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/messages/PrePrepareMsg.hpp:72]: We talked and agreed that the old constructor (without the CID) will call the constructor with the CID (send to_string(seq_num).

File: bftengine/src/bftengine/messages/PrePrepareMsg.hpp (hunk 5 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/messages/PrePrepareMsg.hpp]: Fixed.

File: bftengine/src/bftengine/messages/PrePrepareMsg.hpp (hunk 4 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/messages/PrePrepareMsg.hpp]: Fixed.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (Liorgal28) [reply bftengine/src/bftengine/ReplicaImp.cpp:3612]: Done.

**Review by yuliasherman — APPROVED** [review]
LGTM

**Review by f-squirrel — APPROVED** [review]
LGTM, thank you for addressing the comments!

**Pull request merged**